### PR TITLE
Ticket #2 -- Product Controller with Test

### PR DESCRIPTION
**Pull request by AFranklin41**
## Proposed changes

Product Controller allows the user to `GET` a single entry, and `GET`, `POST`, `PUT`, and `DELETE` entries from the Product table in the BangazonAPI Database. 

Included is the test module `ProductTest` that will test all methods automatically. 

## Related Issues

`Ticket #2 `

https://trello.com/c/xNcjlvgR

## Steps To Test

clone my [repository](https://github.com/NewForce-at-Mountwest/bangazonapi-team-donkoss/tree/af-ticket2)
`git fetch --a` to get access to my branch
`git checkout af-ticket2` switch to my branch
`git pull origin af-ticket2` to grab the latest commit

`start BangazonAPI.csproj` to open the project

create a new database and title it `BangazonAPI`

open a new SQL query in `SQL Object Explorer` and run this [script](https://github.com/NewForce-at-Mountwest/bangazonapi-team-donkoss/blob/61cd0b2c1aed09c6988c016f7dedeae777e0751c/BangazonAPI/dummyData.sql) to populate the database with dummy data

In the Solution Explorer, right click on `BangazonAPITest` and select `Run Tests`

The user should be able to `GET`, `POST`, `PUT`, and `DELETE` entries in the Product table in the BangazonAPI database using Postman.

The controller is set to "soft" delete entries from the database, while the test is set to "hard" delete the dummy data it creates.

## What kind of change does this PR introduce?
Please check one of the following:
- [ ] Bug fix
- [ ] Documentation update
- [x] New feature
- [ ] Other

### CODE REVIEW

**Comment by seaverstanley** [comment]
Tests passed, and the code fetches the right data. Looks good

**Comment by jordan-castelloe** [comment]
Testing!

File: BangazonAPI/Controllers/ProductController.cs [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
+                conn.Open();
+                using (SqlCommand cmd = conn.CreateCommand())
+                {
+                    cmd.CommandText = @"SELECT Product.Id, Product.Price, 
+                                        Product.Title, Product.Description,
+                                        Product.Quantity, Product.Archived,
+                                        Product.CustomerId AS 'Customer Id',
+                                        Product.ProductTypeId AS 'Product Type Id'
+                                        FROM Product 
+                                        LEFT JOIN Customer on Product.CustomerId = Customer.Id
```
Comment (jordan-castelloe) [inline BangazonAPI/Controllers/ProductController.cs]: You don't need these joins-- you're not using any data from these tables. You can just `SELECT` straight from the Product table and make your query a little more efficient 👍

File: BangazonAPI/Controllers/ProductController.cs [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+                    cmd.CommandText = @"INSERT INTO Product (ProductTypeId, CustomerId, Price, Title, Description, Quantity, Archived)
+                                        OUTPUT INSERTED.Id
+                                        VALUES (@productTypeId, @customerId, @price, @title, @description, @quantity, @archived)";
+                    cmd.Parameters.Add(new SqlParameter("@productTypeId", product.ProductTypeId));
+                    cmd.Parameters.Add(new SqlParameter("@customerId", product.CustomerId));
+                    cmd.Parameters.Add(new SqlParameter("@price", product.Price));
+                    cmd.Parameters.Add(new SqlParameter("@title", product.Title));
+                    cmd.Parameters.Add(new SqlParameter("@description", product.Description));
+                    cmd.Parameters.Add(new SqlParameter("@quantity", product.Quantity));
+                    cmd.Parameters.Add(new SqlParameter("@archived", product.Archived));
```
Comment (jordan-castelloe) [inline BangazonAPI/Controllers/ProductController.cs:137]: This is a design decision you should discuss with your group-- do y'all want the end user to be able to create an archived product? Or should `archived` automatically be set to `false` when you create a product, and only changed to `true` if they run a soft delete? (How much do you trust your end user?) I think an argument could be made either way and the client doesn't have a preference about this as long as you can back up your reasoning.

File: BangazonAPI/Controllers/ProductController.cs [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+                    cmd.CommandText = @"SELECT Product.Id, Product.Price, 
+                                        Product.Title, Product.Description,
+                                        Product.Quantity, Product.Archived,
+                                        Product.CustomerId AS 'Customer Id',
+                                        Product.ProductTypeId AS 'Product Type Id'
+                                        FROM Product 
+                                        LEFT JOIN Customer on Product.CustomerId = Customer.Id
+                                        LEFT JOIN ProductType on Product.ProductTypeId = ProductType.Id";
+
+                    SqlDataReader reader = cmd.ExecuteReader();
```
Comment (jordan-castelloe) [inline BangazonAPI/Controllers/ProductController.cs]: With soft delete-- products that are archived _shouldn't_ come back when you get the full list. It's okay if you support a singular `GET` for archived products, but they shouldn't appear in the `GET` all.

**Review by jordan-castelloe — COMMENTED** [review]
I made some small suggestions-- you're super close. Good job! I'd like you to at least change it so that archived products don't show up when you `GET` a list of all the products so that soft delete emulates real delete functionality. My other suggestions are more about design/ efficiency than functionality. I don't need to re-approve this if you make changes-- consider this a 👍 pending the `GET` all fix.

**Pull request merged**